File: python/worker.py
from utils import *
import asyncio
from metrics import Metrics
from dynamodb_service import DynamoDBService
import time


class Worker:
  def __init__(self, partitions):
    self.partitions = partitions
    self.metrics = Metrics(time.time()*1000, partitions)
    self.ddb_service = DynamoDBService(self.metrics)
    self.scan_times = dict()
  
  async def start(self):
    # sys.setrecursionlimit(10000)
    now = time.time() * 1000
    for partition in self.partitions:
      self.scan_times[partition] = now

    async with self.ddb_service.ddb_client as ddb_client:
      self.table = await ddb_client.Table(self.ddb_service.table_name)
      await self.scan_group(self.partitions)
  
  async def scan_group(self, partitions):
    start = time.time() * 1000
    no_delay = False
    
    # iterate partitions
    for partition in partitions:
      now = time.time() * 1000
      if(self.scan_times[partition] <= now):
        schedule_immediate = await self.dispatch_overdue(partition)
        next_scan = time.time() * 1000 if schedule_immediate else (time.time() + 1) * 1000
        self.scan_times[partition] = next_scan
        no_delay = no_delay or schedule_immediate
      else:
        no_delay = False
    self.metrics.scan_group(time.time()*1000-start, partitions, no_delay)

    # next scan
    if not no_delay:
      await asyncio.sleep(1)
    else:
      await asyncio.sleep(0)     # maximum recursion depth exceeded 에러 때문에 추가
    await self.scan_group(partitions)

    
  async def dispatch_overdue(self, partition) -> bool:
    start = time.time() * 1000
    try:
      schedule_query_response = await self.ddb_service.get_overdue_jobs(self.table, partition)
      schedules = schedule_query_response.schedules
      async with asyncio.TaskGroup() as tg:
        for schedule in schedules:
          tg.create_task(self.after_dispatch(schedule))
      print(f'fetch done from partition {partition}')
      self.metrics.dispatch_overdue(time.time()*1000-start, partition, len(schedules))
      return schedule_query_response.should_immediately_query_again
    except Exception as err:
      print(err)
      return True


  async def after_dispatch(self, schedule):
    schedule = await self.ddb_service.update_status(self.table, schedule, 'SCHEDULED', 'ACQUIRED')
    await self.dispatch_to_destination(schedule)


  async def dispatch_to_destination(self, schedule):
    message = f'''
== PUSH JOB TO DESTINATION QUEUE ==
QUEUE: {schedule.job_spec['queueName']}
JOB: {schedule.job_spec['jobClass']}
PARAMS: {schedule.job_spec['jobParams']}
'''
    print(message)
    await self.ddb_service.delete_dispatched_job(self.table, schedule)


async def test():
  await Worker([1,2,3,4,5,6,7,8]).start()


if __name__ == "__main__":
  asyncio.run(test())

File: ruby/worker.rb
require_relative './dynamodb_service'
require_relative './metrics'

class Worker
  include Concurrent::Promises::FactoryMethods

  def initialize(partitions)
    @metrics = Metrics.new(partitions)
    @partitions = partitions
    @ddbsvc = DynamoDBService.new(@metrics)
    @scan_times = Hash.new
    current_time = Time.now
    @partitions.each do |partition|
      @scan_times[partition] = current_time
    end
  end

  # return value : future<nil>
  def dispatch_to_destination(schedule)
    msg = 
    puts "== PUSH JOB TO DESTINATION QUEUE =="
    puts "QUEUE: #{schedule.job_spec['queueName']}"
    puts "JOB: #{schedule.job_spec['jobClass']}"
    puts "QUEUE: #{schedule.job_spec['jobParams']}"
    puts ''
    @ddbsvc.delete_dispatched_jobs(schedule)
  end

  # return value : future<bool>
  def dispatch_overdue(partition)
    start = Time.now
    done = resolvable_future
    @ddbsvc.get_overdue_jobs(partition)
    .then (done) do |query_result, done|
      tasks = query_result[:schedules].map do |schedule|
        @ddbsvc.update_status(schedule, 'SCHEDULED', 'ACQUIRED')
        .then {|schedule| dispatch_to_destination(schedule)}
      end
      zip(*tasks).value!
      @metrics.dispatch_overdue(Time.now-start, partition, query_result[:schedules].count)
      done.fulfill query_result[:should_immediately_query_again]
    end
    return done
  end

  def scan_group
    start = Time.now
    puts 'start scanning...'
    no_delay = false
    @partitions.each do |partition|
      now = Time.now
      if @scan_times[partition] <= now
        this_no_delay = dispatch_overdue(partition)
        .then do |scanImmediate|
          next_scan = scanImmediate ? Time.now : Time.now + 1
          @scan_times[partition] = next_scan
          scanImmediate
        end.value!
        no_delay ||= this_no_delay
      else
        no_delay = false
      end
    end

    @metrics.scan_group(Time.now-start, no_delay)
    sleep(1) unless no_delay
    scan_group
  end
end